runtime/Cpp/runtime/src/atn/ATNStateType.h

```diff
@@ -0,0 +1,36 @@
+/* Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
+ * Use of this file is governed by the BSD 3-clause license that
+ * can be found in the LICENSE.txt file in the project root.
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <string>
+
+#include "antlr4-common.h"
+
+namespace antlr4 {
+namespace atn {
+
+  // Constants for ATNState serialization.
+  enum class ATNStateType : size_t {
+    INVALID = 0,
+    BASIC = 1,
+    RULE_START = 2,
+    BLOCK_START = 3,
+    PLUS_BLOCK_START = 4,
+    STAR_BLOCK_START = 5,
+    TOKEN_START = 6,
+    RULE_STOP = 7,
+    BLOCK_END = 8,
+    STAR_LOOP_BACK = 9,
+    STAR_LOOP_ENTRY = 10,
+    PLUS_LOOP_BACK = 11,
+    LOOP_END = 12,
+  };
+
+  std::string atnStateTypeName(ATNStateType atnStateType);
```

Comment:
Not necessarily now (not sure what the current baseline requirements are, is antlr already compiling with c++17?), but longer term, using `std::string_view` here is probably a good idea.

Reply:
We could use string_view, but then we cannot ever dynamically allocate. For example, if we wanted to do "INVALID(%d)".

Reply:
Given that `ATNStateType` is an enum class it would be very hard to get an invalid value in there. So I'd prefer performance over new string allocations to be prepared for an unlikely force-typecast any time.

Reply:
atnStateTypeName and transitionTypeName are only used by toString methods, which are not called anywhere in the ANTLR runtime except for debugging, which is off by default. So performance is not an issue in this case, unless the user is calling toString in which cause I would suggest not unless debugging because a bunch of them are not performant.